### PR TITLE
add slave processes support

### PR DESCRIPTION
**Pull request by luckydrq**
PR for [#12](https://github.com/node-modules/cfork/issues/12)

### CODE REVIEW

**Comment by luckydrq** [comment]
@fengmk2 please review this PR to see if it is OK

File: index.js [hunk 1]
```diff
@@ -93,7 +94,7 @@ function fork(options) {
     }
     disconnects[worker.process.pid] = new Date();
     if (allow()) {
-      cluster.fork();
+      forkWorker(worker);
```
Comment (fengmk2) [inline index.js]: worker 会有 `_clusterSettings` 属性？

File: index.js [hunk 2]
```diff
@@ -177,4 +191,32 @@ function fork(options) {
...
+      return null;
+    } else {
+      return opt;
+    }
+  }
+
+  /**
+   * fork worker with certain settings
+   */
+  function forkWorker(worker) {
```
Comment (fengmk2) [inline index.js]: 应该 `forkWorker(settings)` ，这样比较好，forkWorker 不应该依赖 worker 的私有属性 _clusterSettings 的。

File: index.js (hunk 1 above)
Comment (luckydrq) [inline index.js]: 这个属性是我加的，在`worker`上保存一下特定的`cluster.settings`。因为slave进程也采用`cluster.fork()`来生成，即也是一个`worker`，但与其他`worker`的区别是`settings`不一样（`exec`, `args`）。所以当`refork`某个`worker`的时候需要调用`worker`自己的`settings`。

File: index.js (hunk 2 above)
Comment (luckydrq) [inline index.js]: 确实，我改一下这里

**Comment by luckydrq** [comment]
@fengmk2 i've done some modifications according to your suggestions and re-commit.

**Comment by luckydrq** [comment]
Any news?

File: index.js [hunk 3]
```diff
@@ -93,7 +95,8 @@ function fork(options) {
     }
     disconnects[worker.process.pid] = new Date();
     if (allow()) {
-      cluster.fork();
+      newWorker = forkWorker(worker._clusterSettings);
+      newWorker._clusterSettings = worker._clusterSettings;
```
Comment (fengmk2) [inline index.js]: why worker need `_clusterSettings` property?

Comment (fengmk2) [inline index.js]: Oh, I see, worker die and retry.

Comment (fengmk2) [inline index.js]: Has a test for this case? Slave dead and refork again?

File: test/cfork.test.js [hunk 4]
```diff
@@ -101,4 +105,27 @@ describe('cfork.test.js', function () {
...
+      should.not.exist(err);
+      body.toString().should.equal('GET /');
+      res.statusCode.should.equal(200);
+      done();
+    });
+  });
+
+  it('should mock slave error', function (done) {
+    urllib.request('http://localhost:1985/error', function (err) {
+      should.exist(err);
```
Comment (fengmk2) [inline test/cfork.test.js]: request normal url again to make sure slave refork work.

File: test/cfork.test.js [hunk 5]
```diff
@@ -66,6 +70,34 @@ describe('cfork.test.js', function () {
...
+      urllib.request('http://localhost:1985/', function (err, body, res) {
+        should.not.exist(err);
+        body.toString().should.equal('GET /');
+        res.statusCode.should.equal(200);
+        done();
+      });
+    });
+  });
+
```
Comment (luckydrq) [inline test/cfork.test.js]: Slave die and refork test case added. @fengmk2

**Comment by fengmk2** [comment]
+1

**Comment by fengmk2** [comment]
I will publish when I get the mbp

**Comment by fengmk2** [comment]
1.4.0

**Pull request merged**